AttributeError: 'NeuralNetClassifier' object has no attribute 'score'

damienlancry opened the pull request:
These changes make NeuralNetClassifier NeuralNetBinaryClassifier and NeuralNetRegressor inherit from sklearn.base.ClassifierMixin and sklearn.base.RegressorMixin to gives them a score method. While making these changes I encountered the error ```Cannot import 'NeuralNetBinaryClassifier'``` so i added it to the ```__init__``` file.  This is in response to the issue #469  that I opened.

Here are working examples showing the efficiency of my changes:
```
#### Classification

from sklearn.model_selection import train_test_split
from skorch import NeuralNetClassifier 
from torch import nn, optim, tensor, cuda, Tensor
import torch.nn.functional as F 
from skorch.dataset import Dataset
from skorch.helper import predefined_split
from sklearn.datasets import load_digits, load_breast_cancer


digits = load_digits()
X_train, y_train = digits.data, digits.target
X_train = X_train.reshape((-1, 1, 8, 8)) / 16.
X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size=0.2)

class ClassifierModule(nn.Module):
    def __init__(self):
        super(ClassifierModule, self).__init__()
        self.layer1 = nn.Sequential(
        nn.Conv2d(1, 32, 3),
        nn.ReLU(),
        nn.Conv2d(32, 32, 3),
        nn.ReLU(),
        )
        self.d1 = nn.Dropout(0.25)
        self.f1 = nn.Linear(4 * 4 * 32, 64)
        self.d2 = nn.Dropout(0.5)
        self.f2 = nn.Linear(64, 10)
    def forward(self, x):
        out = x
        out = self.layer1(out)
        out = out.view(-1,4 * 4 * 32)
        out = self.d1(out)
        out = self.f1(out)
        out = self.d2(out)
        out = self.f2(out)
        return F.softmax(out, dim=-1)

def tensorize_feats(*features):
    for array in list(features):
        yield tensor(array).float().cuda()
def tensorize_labels(*labels):
    for array in list(labels):
        yield tensor(array).long().cuda()
        
X_train, X_test = tensorize_feats(*(X_train, X_test))
y_train, y_test = tensorize_labels(*(y_train, y_test))

model = NeuralNetClassifier(ClassifierModule, 
                            max_epochs=100, 
                            batch_size=128, 
                            lr=0.001, 
                            device="cuda",
                            optimizer=optim.Adam, 
                            verbose=0
                           )

model.fit(X_train, y_train)
model.score(X_test.cpu(),y_test.cpu())

### Binary Classification

from skorch import NeuralNetBinaryClassifier 

digits = load_digits()
X, y = digits.data, digits.target
idx = np.where(y < 2)[0]
X, y = X[idx], y[idx] 
X = X.reshape((-1, 1, 8, 8)) / 16.



X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

class ClassifierModule(nn.Module):
    def __init__(self):
        super(ClassifierModule, self).__init__()
        self.layer1 = nn.Sequential(
        nn.Conv2d(1, 16, 3),
        nn.ReLU(),
        nn.Conv2d(16, 16, 3),
        nn.ReLU(),
        )
        self.d1 = nn.Dropout(0.25)
        self.f1 = nn.Linear(4 * 4 * 16, 32)
        self.d2 = nn.Dropout(0.5)
        self.f2 = nn.Linear(32, 1)
    def forward(self, x):
        out = x
        out = self.layer1(out)
        out = out.view(-1,4 * 4 * 16)
        out = self.d1(out)
        out = self.f1(out)
        out = self.d2(out)
        out = self.f2(out)
        return out.view(-1)

def tensorize(*features):
    for array in list(features):
        yield tensor(array).float().cuda()
        
X_train, X_test = tensorize(*(X_train, X_test))
y_train, y_test = tensorize(*(y_train, y_test))

model = NeuralNetBinaryClassifier(ClassifierModule, 
                            max_epochs=100, 
                            batch_size=128, 
                            lr=0.001, 
                            device="cuda",
                            optimizer=optim.Adam,
                            verbose=0
                           )

model.fit(X_train, y_train)
model.score(X_test.cpu(),y_test.cpu())

### Regression

from skorch import NeuralNetRegressor

digits = load_digits()
X, y = digits.data, digits.target
X = X.reshape((-1, 1, 8, 8)) / 16.
y = y.reshape((-1, 1)) / 10.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)


class RegressorModule(nn.Module):
    def __init__(self):
        super(RegressorModule, self).__init__()
        self.layer1 = nn.Sequential(
        nn.Conv2d(1, 16, 3),
        nn.ReLU(),
        nn.Conv2d(16, 16, 3),
        nn.ReLU(),
        )
        self.d1 = nn.Dropout(0.25)
        self.f1 = nn.Linear(4 * 4 * 16, 32)
        self.d2 = nn.Dropout(0.5)
        self.f2 = nn.Linear(32, 1)
    def forward(self, x):
        out = x
        out = self.layer1(out)
        out = out.view(-1,4 * 4 * 16)
        out = self.d1(out)
        out = self.f1(out)
        out = self.d2(out)
        out = self.f2(out)
        return out

def tensorize(*features):
    for array in list(features):
        yield tensor(array).float().cuda()
        
X_train, X_test = tensorize(*(X_train, X_test))
y_train, y_test = tensorize(*(y_train, y_test))

model = NeuralNetRegressor(RegressorModule, 
                            max_epochs=100, 
                            batch_size=128, 
                            lr=0.001, 
                            device="cuda",
                            optimizer=optim.Adam,
                            verbose=1
                          )

model.fit(X_train, y_train)
model.score(X_test.cpu(),y_test.cpu())
```

BenjaminBossan commented:
Thank you for working on this.

Could you please
- [ ] write tests for each case (for regression, please also test multioutput).
- [ ] update the docs (after the [section for predict/predict_proba](https://github.com/skorch-dev/skorch/blob/82efd5a94b2477c5d8e4be8c3cd306f1504df453/docs/user/neuralnet.rst#predictx-and-predict_probax))
- [ ] update the CHANGES.md.

damienlancry commented:
Thank you for considering these changes.

Should I write tests in a specific file? 
Are the examples i gave for classification, binary classification and regression sufficient tests?

BenjaminBossan commented:
> Should I write tests in a specific file?

Yes, please add them to `test_classifier.py` and `test_regressor.py`.

> Are the examples i gave for classification, binary classification and regression sufficient tests?

I would say that these examples are a little bit too much even (from the point of view of LOCs and training time). Please consider using the existing fixtures for data and nets. E.g., look at this test here:

https://github.com/skorch-dev/skorch/blob/82efd5a94b2477c5d8e4be8c3cd306f1504df453/skorch/tests/test_classifier.py#L58-L62

You can re-use the same net and data but instead of testing `predict_proba`, you would test `score`. You don't even need to test whether the score is good, just that its values conform to expectation (e.g. 0 <= p <= 1).

For the regression, I believe sklearn should be able to handle multioutput. Maybe you could add a minimal test for this as well. If you are unsure about something, just ask.

damienlancry commented:
ok one more question: I have no idea what dataset to use for the multi output regression

BenjaminBossan commented:
You could use `sklearn.dataset.make_regression(..., n_targets=3)` for multiregression. As for the module, you could use `skorch.toy.MLPModule` with `output_units=3`.

damienlancry commented:
Hi
I added tests, updated the changes.md and the docs.
Here are simpler working examples (I added one for the multi target regression case):
```
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_classification, make_regression
from skorch.toy import make_classifier, make_binary_classifier, make_regressor
from skorch import NeuralNetClassifier, NeuralNetBinaryClassifier, NeuralNetRegressor
from torch import tensor, optim, nn

###################### CLASSIFICATION ###############################
n_classes  = 3
n_features = 5
X, y = make_classification(n_samples=10000, n_features=n_features, n_informative=3, n_classes=n_classes)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
X_train, X_test = tensor(X_train).float().cuda(), tensor(X_test).float()
y_train, y_test = tensor(y_train).long().cuda(),  tensor(y_test).long()

Net  = make_classifier(output_nonlin=None, input_units=n_features, output_units=n_classes, num_hidden=2)
model= NeuralNetClassifier(Net, device="cuda", optimizer=optim.Adam, criterion=nn.CrossEntropyLoss)

model.fit(X_train, y_train)
print(model.score(X_test,y_test))

###################### BINARY CLASSIFICATION #######################
n_classes  = 2
n_features = 5
X, y = make_classification(n_samples=10000, n_features=n_features, n_informative=3, n_classes=n_classes)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
X_train, X_test = tensor(X_train).float().cuda(), tensor(X_test).float()
y_train, y_test = tensor(y_train).float().cuda(), tensor(y_test).float()

Net  = make_binary_classifier(input_units=n_features, num_hidden=2, output_units=1)
model= NeuralNetBinaryClassifier(Net, device="cuda",optimizer=optim.Adam, criterion=nn.BCEWithLogitsLoss)

model.fit(X_train, y_train)
print(model.score(X_test, y_test))

###################### REGRESSION ##################################
n_features = 5
X, y = make_regression(n_samples=10000, n_features=n_features)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
X_train, X_test = tensor(X_train).float().cuda(), tensor(X_test).float()
y_train, y_test = tensor(y_train).float().cuda(), tensor(y_test).float()

y_train, y_test = y_train.view(-1, 1), y_test.view(-1, 1)

Net  = make_regressor(input_units=n_features)
model= NeuralNetRegressor(Net, device="cuda", optimizer=optim.Adam)

model.fit(X_train, y_train)
print(model.score(X_test, y_test))

###################### MULTI TARGET REGRESSION #####################
n_features = 5
n_targets  = 3
X, y = make_regression(n_samples=10000, n_features=n_features, n_targets=n_targets)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
X_train, X_test = tensor(X_train).float().cuda(), tensor(X_test).float()
y_train, y_test = tensor(y_train).float().cuda(), tensor(y_test).float()

y_train, y_test = y_train.view(-1, n_targets), y_test.view(-1, n_targets)

Net  = make_regressor(input_units=n_features, output_units=n_targets)
model= NeuralNetRegressor(Net, device="cuda", optimizer=optim.Adam)

model.fit(X_train, y_train)
print(model.score(X_test, y_test))
```

damienlancry commented:
It seems that "All checks have failed" and I have no idea why. I am not used to contributing to open source projects like that, so I do not really where to look to diagnose the problem. Any suggestion?

BenjaminBossan commented:
Thanks for the contribution, I will review this soon. Regarding the failing CI, this is unrelated to your changes, so don't worry about it.

BenjaminBossan commented:
And I forgot to mention: For the CI to pass (and presumably your local tests too), you need to either merge the current master or rebase on it.

BenjaminBossan commented on docs/user/neuralnet.rst:
```diff
@@ -359,7 +359,8 @@ score(X, y)
 This method returns the mean accuracy on the given test data and labels for
 classifiers and the coefficient of determination R^2 of the prediction for
-regressors.
+regressors. NeuralNet Class still has no score method. If one needs it,
```
Could you please replace "NeuralNet Class" with ":class:\`.NeuralNetClassifier\`", this should create the correct link in the docs.

BenjaminBossan commented on docs/user/neuralnet.rst:
```diff
@@ -359,7 +359,8 @@ score(X, y)
 classifiers and the coefficient of determination R^2 of the prediction for
-regressors.
+regressors. NeuralNet Class still has no score method. If one needs it,
+one needs to implement it himself.
```
We prefer to speak to the user directly in the docs: "If you need it, you have to implement it yourself."

BenjaminBossan commented on skorch/tests/test_regressor.py:
```diff
@@ -90,4 +107,8 @@ def test_score(self, net, data):
         assert r2_score <= 1.
-        return r2_score
+
+    def multi_target_test_score(self, multi_target_net, multi_target_data):
```
All tests have to be named `test_*` for pytest to discover them. You need to rename the test to something like `test_multi_target_score`.

BenjaminBossan commented on skorch/tests/conftest.py:
```diff
@@ -78,6 +78,15 @@ def regression_data():
+def multi_target_regression_data():
+    X, y = make_regression(
+        1000, 20, n_targets=3, n_informative=10, bias=0, random_state=0)
+    X, y = X.astype(np.float32), y.astype(np.float32).reshape(-1, 1)
```
You should not reshape the `y` data to `(-1, 1)`. This will make it a 1-dimensional array, defeating the purpose of multioutput regression.

BenjaminBossan commented on skorch/tests/conftest.py:
```diff
@@ -78,6 +78,15 @@ def regression_data():
     return Xt, yt
 
+@pytest.fixture(scope='module')
+def multi_target_regression_data():
```
In sklearn lingo, this is called `multioutput`, not `multitarget`

BenjaminBossan commented on skorch/tests/test_regressor.py:
```diff
@@ -36,6 +36,23 @@ def net(self, net_cls, module_cls):
         )
 
+    @pytest.fixture(scope='module')
+    def multi_target_data(self, multi_target_regression_data):
```
Since this fixture doesn't really do anything, you can remove it. Below, in the test function that uses the data, you can directly reference the fixture you defined in `conftest.py`.

BenjaminBossan commented on skorch/tests/test_regressor.py:
```diff
@@ -90,4 +107,8 @@ def test_score(self, net, data):
+
+    def multi_target_test_score(self, multi_target_net, multi_target_data):
+        X, y = multi_target_data
+        r2_score = multi_target_net.score(X, y)
```
Before scoring, you still need to call `fit`.

damienlancry replied:
Ok but I did not do it for the single output regression and for the classification. Should I do it as well?

BenjaminBossan replied:
No need, the other nets are already fitted at this point, only this one isn't.

BenjaminBossan commented on skorch/tests/test_regressor.py:
```diff
@@ -36,6 +36,23 @@ def net(self, net_cls, module_cls):
+    def multi_target_net(self, net_cls, multi_target_module_cls):
+        return net_cls(
+            multi_target_module_cls,
+            max_epochs=20,
```
Please only use 1 epoch. For the test, it doesn't matter how well the model is fit, but the test time will increase considerably with 20 epochs.

damienlancry replied:
Ok but in your ```net``` function (single output regression) you used ```max_epochs=20```. Should I fix this?

damienlancry commented:
Again the checks are failing, I am not sure if this is because I messed up when pushing my changes or if this is because there really is a failing test... Sorry about that

BenjaminBossan commented:
Hey, no worries, this is why there are automatic tests.

The error is that you want to use the fixture `multioutput_data`, but in conftest it is called `multioutput_regression_data`. You should probably use the latter name, since it's more descriptive.

In general, you can also run the tests locally before pushing to catch any such errors. To do that, follow the instructions for developers in the [installation section](https://github.com/skorch-dev/skorch#installation) in the README. Or you could click on `Details` link next to the failing Travis CI check and inspect the error there.

damienlancry commented:
ok thanks for the instructions I think I fixed the remaining problems now :)

BenjaminBossan commented on skorch/tests/conftest.py:
```diff
@@ -78,6 +78,15 @@ def regression_data():
+def multioutput_regression_data():
+    X, y = make_regression(
+        1000, 20, n_targets=3, n_informative=10, bias=0, random_state=0)
+    X, y = X.astype(np.float32), y.astype(np.float32).reshape(-1, 3)
```
The `.reshape(-1, 3)` should be unnecessary.

damienlancry replied:
Yeah right

BenjaminBossan commented on docs/user/neuralnet.rst:
```diff
@@ -354,6 +354,14 @@ Obviously, this only makes sense if
+
+This method returns the mean accuracy on the given test data and labels for
+classifiers and the coefficient of determination R^2 of the prediction for
+regressors. :class:`.NeuralNet` Class still has no score method. If you need it,
```
`Class` is not necessary.

thomasjpfan commented on docs/user/neuralnet.rst:
```diff
@@ -354,6 +354,14 @@ Obviously, this only makes sense if
+score(X, y)
+^^^^^^^^^^^
+
+This method returns the mean accuracy on the given test data and labels for
```
`X` could be test or training data.

```suggestion
This method returns the mean accuracy on the given data and labels for
```

damienlancry replied:
True True True

damienlancry commented:
How about this last commit?

BenjaminBossan commented:
Great, thank you for incorporating all the requested changes.
